Extensions/PathfindingBehavior/pathfindingobstacleruntimebehavior.ts

```diff
@@ -190,7 +197,38 @@ namespace gdjs {
     doStepPostEvents(runtimeScene: gdjs.RuntimeScene) {}
 
     getAABB() {
-      return this.owner.getAABB();
+      if (!this._hitBoxesAABBUpToDate) {
```

Comment:
Can you tell more about this? In which situations this caching works and what's the idea behind the invalidation of the cache?

I see a `this._hitBoxesAABBUpToDate = false` in pathfindingruntimebehavior but not in this file, which would indicate that the hitboxes are never invalidated?

More generally, I wonder if we can avoid transforming the AABB at all, because I see a lot of risk of having them de-synced with the rest (object changing, the isometry viewpoint changing...) and instead do the pathfinding search in a way that each position considered is following the isometry transformation?

Reply:
Yes, I forgot to copy/paste some lines.
The thing is that the transformation of the AABB is not the same as the AABB of the transformed hitbox plygon.

Reply:
Do you think that if, instead of AABB, we used more precise methods to check if a point is inside a polygon (i.e: if we use object hitboxes), then we could avoid any "transformation" of AABB at all, and we could just apply a different distance function to the algorithm, and a different X/Y position transformation to nodes?
So that we go from the left hand side diagram to the right one:

![image](https://user-images.githubusercontent.com/1280130/107334560-26f3bf80-6ab7-11eb-95d5-cce34b197f08.png)

This would make the search to be following the "isometry" viewpoint like seen on the screen, without transforming AABB/hitboxes of objects.

This means that objects hitboxes, would be already defined in the editor like this: 
![image](https://user-images.githubusercontent.com/1280130/107334852-894cc000-6ab7-11eb-9104-f480d48ca1c6.png)
(which is what is done in the isometry example, using big red rectangle obstacles: 
![image](https://user-images.githubusercontent.com/1280130/107334957-a7b2bb80-6ab7-11eb-818c-9b4a24bc102e.png)
 )
 
 Not sure if it makes sense, let me know :) Being able to search "naturally" in a viewpoint like this with existing polygons would reduce the number of transformations to do I think, and would work directly with hitboxes as long as they are well positioned already like the ones I shown.

Reply:
Yep - note that I'm not sure how easy/hard it will be rework the behavior to use hitboxes rather than just AABB (and we might want to keep a boolean to use just AABB, to keep compatibility with existing games and improve performance in cases where hitboxes are too costly to use).
But it would be great to do, because I think it has confused users already, and it would open the way to this isometry-enabled pathfinding by "just" exploring nodes using different positions/distance function.

Reply:
> 
> 
> Yep - note that I'm not sure how easy/hard it will be rework the behavior to use hitboxes rather than just AABB (and we might want to keep a boolean to use just AABB, to keep compatibility with existing games and improve performance in cases where hitboxes are too costly to use).


Yes, I think the boolean parameter is a must because the margin parameter can't be easily used with hitboxes so the exact collision mode will not behave the same even for default hitboxes.

Reply:
Yeah exactly. Though this approach can also create another problem. A 1px space between two obstacles could or could not be considered as an open node (i.e: a path to explore) according to if the virtual nodes are exactly positioned between them or not.

Not sure how important this is, but I think this problem was less common in the approach with AABB because even if you add/remove some offset to your virtual grid position, it won't in most cases open new "paths" between obstacles, because if an AABB is even slightly on a virtual cell (i.e: a node), then the node is considered as having the obstacle on it.

Not sure what's the right solution yet. Should we do it the other way, don't check if the center of the node is inside an obstacle, but rather check for every point of the obstacle hitbox if it's inside the node. But not sure how complex this is.

Reply:
I'm not sure to follow.
My understanding is that the shape of the object is stamped on each cell to test the collision with surroundings obstacles.
If we use the hitboxes instead of the AABB as the shape. I don't see why interstices of 1px could be an issue.
Do you mean that using the node as the shape could be an option? I'm not sure to see use cases.

About the performance issue, if hitboxes collision is too heavy, I wonder if RuntimeObject could have a isAABBMatchingHitboxes() to allow optimizations because I think it is commonly the case.
Maybe, inner AABB (I don't know the right term) could be interesting to ensure collisions the same way outer AABB discredit it.